### PR TITLE
Adds middleware that generates cProfile profile for a request

### PR DESCRIPTION
**Pull request by dkliban**
The request needs to have a X-PROFILE-REQUEST header for the middleware to generate a profile. The middleware logs a URL that can be used to download the profile.

### CODE REVIEW

File: pulp_service/pulp_service/app/middleware.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    PROFILER_REQUEST_ATTR_NAME = '_django_cprofile_middleware_profiler'
+
+    def can(self, request):
+        if 'HTTP_X_PROFILE_REQUEST' in request.META:
+            return True
+        return False
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        if self.can(request):
+            profiler = cProfile.Profile()
```
Comment (bmbouter) [inline pulp_service/pulp_service/app/middleware.py:31]: We have this dependency already yes?

File: deploy/clowdapp.yaml [hunk 2]
```diff
@@ -826,3 +828,6 @@ parameters:
   - name: PULP_STATIC_URL
     description: The path to the static assets needed by some Django apps
     value: "/api/pulp/assets/"
+  - name: PULP_MIDDLEWARE
+    description: List of Middleware
+    value: "@merge pulp_service.app.middleware.ProfilerMiddleware"
```
Comment (bmbouter) [inline deploy/clowdapp.yaml:833]: Are we sure this will be earliest in the middleware?

File: pulp_service/pulp_service/app/middleware.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    Simple profile middleware to profile django views. To run it, add
+    x-profile-request header to the request.
+
+    This is adapted from an example found here:
+    https://github.com/omarish/django-cprofile-middleware/blob/master/django_cprofile_middleware/middleware.py
+    """
+    PROFILER_REQUEST_ATTR_NAME = '_django_cprofile_middleware_profiler'
+
+    def can(self, request):
+        if 'HTTP_X_PROFILE_REQUEST' in request.META:
```
Comment (bmbouter) [inline pulp_service/pulp_service/app/middleware.py:26]: So the first draft is entirely yes/no based on the presence of the header? I'm ok with that if so.

File: pulp_service/pulp_service/app/middleware.py (hunk 1 above)
Comment (dkliban) [reply pulp_service/pulp_service/app/middleware.py:31]: it's part of the standard library https://docs.python.org/3.8/library/profile.html

File: deploy/clowdapp.yaml (hunk 2 above)
Comment (dkliban) [reply deploy/clowdapp.yaml:833]: It will be the last middleware in the list.

File: pulp_service/pulp_service/app/middleware.py (hunk 3 above)
Comment (dkliban) [reply pulp_service/pulp_service/app/middleware.py:26]: that is correct

File: deploy/clowdapp.yaml (hunk 2 above)
Comment (bmbouter) [reply deploy/clowdapp.yaml:833]: That's better than not in the list! Is there a way to make it first? I'm interested in investigating the basic auth middleware overhead for example. If not I'm ok with trying this as-is also.

**Pull request merged**